File: backend/app/schemas/income.py
# backend/app/schemas/income.py
"""
Income Schemas
Pydantic models for income sources and history API
"""

from pydantic import BaseModel, Field, field_validator
from typing import Optional, List
from uuid import UUID
from datetime import date, datetime
from decimal import Decimal
from enum import Enum


# ============================================
# ENUMS
# ============================================

class IncomeSourceType(str, Enum):
    JOB = "job"
    SCHOLARSHIP = "scholarship"
    STIPEND = "stipend"
    FREELANCE = "freelance"
    INVESTMENT = "investment"
    GIFT = "gift"
    OTHER = "other"


class PayStructure(str, Enum):
    HOURLY = "hourly"
    SALARY = "salary"
    FIXED = "fixed"
    PER_SESSION = "per_session"
    COMMISSION = "commission"


class PaymentFrequency(str, Enum):
    WEEKLY = "weekly"
    BIWEEKLY = "biweekly"
    MONTHLY = "monthly"
    SEMESTER = "semester"
    QUARTERLY = "quarterly"
    ANNUAL = "annual"
    ONE_TIME = "one_time"


class TaxWithholdingType(str, Enum):
    W2 = "w2"
    W1099 = "1099"
    NONE = "none"


# ============================================
# CREATE/UPDATE SCHEMAS
# ============================================

class IncomeSourceCreate(BaseModel):
    """Create a new income source"""
    source_name: str = Field(..., min_length=1, max_length=255)
    source_type: IncomeSourceType
    employer_name: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    
    # Pay structure
    pay_structure: PayStructure
    pay_rate: Optional[Decimal] = Field(None, ge=0)
    pay_unit: Optional[str] = Field(None, max_length=20)
    fixed_amount: Optional[Decimal] = Field(None, ge=0)
    
    # Frequency
    frequency: PaymentFrequency
    next_payment_date: Optional[date] = None
    
    # Work constraints
    max_hours_per_week: Optional[int] = Field(None, ge=1, le=168)
    expected_hours_per_period: Optional[Decimal] = Field(None, ge=0)
    
    # Tax
    is_taxable: bool = True
    tax_rate_federal: Optional[Decimal] = Field(None, ge=0, le=100)
    tax_rate_state: Optional[Decimal] = Field(None, ge=0, le=100)
    tax_rate_local: Optional[Decimal] = Field(None, ge=0, le=100)
    tax_rate_fica: Optional[Decimal] = Field(7.65, ge=0, le=100)
    tax_withholding_type: TaxWithholdingType = TaxWithholdingType.W2
    
    # Reliability
    is_guaranteed: bool = True
    reliability_score: Decimal = Field(1.0, ge=0, le=1)
    
    # Dates
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    
    notes: Optional[str] = None
    
    @field_validator('pay_rate', 'fixed_amount')
    @classmethod
    def validate_pay_info(cls, v, info):
        # At least one of pay_rate or fixed_amount should be provided
        return v


class IncomeSourceUpdate(BaseModel):
    """Update an income source"""
    source_name: Optional[str] = Field(None, min_length=1, max_length=255)
    source_type: Optional[IncomeSourceType] = None
    employer_name: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    
    pay_structure: Optional[PayStructure] = None
    pay_rate: Optional[Decimal] = Field(None, ge=0)
    pay_unit: Optional[str] = Field(None, max_length=20)
    fixed_amount: Optional[Decimal] = Field(None, ge=0)
    
    frequency: Optional[PaymentFrequency] = None
    next_payment_date: Optional[date] = None
    
    max_hours_per_week: Optional[int] = Field(None, ge=1, le=168)
    expected_hours_per_period: Optional[Decimal] = Field(None, ge=0)
    
    is_taxable: Optional[bool] = None
    tax_rate_federal: Optional[Decimal] = Field(None, ge=0, le=100)
    tax_rate_state: Optional[Decimal] = Field(None, ge=0, le=100)
    tax_rate_local: Optional[Decimal] = Field(None, ge=0, le=100)
    tax_rate_fica: Optional[Decimal] = Field(None, ge=0, le=100)
    tax_withholding_type: Optional[TaxWithholdingType] = None
    
    is_guaranteed: Optional[bool] = None
    reliability_score: Optional[Decimal] = Field(None, ge=0, le=1)
    
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    
    notes: Optional[str] = None
    is_active: Optional[bool] = None


# ============================================
# INCOME LOGGING SCHEMAS
# ============================================

class LogIncomeRequest(BaseModel):
    """Log actual income received"""
    # Work done (for hourly/session)
    hours_worked: Optional[Decimal] = Field(None, ge=0)
    days_worked: Optional[Decimal] = Field(None, ge=0)
    sessions_worked: Optional[int] = Field(None, ge=0)
    
    # Amounts
    gross_amount: Decimal = Field(..., ge=0)
    
    # Tax deductions (optional - can be calculated)
    tax_federal: Optional[Decimal] = Field(None, ge=0)
    tax_state: Optional[Decimal] = Field(None, ge=0)
    tax_local: Optional[Decimal] = Field(None, ge=0)
    tax_fica: Optional[Decimal] = Field(None, ge=0)
    other_deductions: Optional[Decimal] = Field(None, ge=0)
    
    net_amount: Optional[Decimal] = Field(None, ge=0)  # Can be calculated
    
    # Payment details
    payment_date: date
    payment_period_start: Optional[date] = None
    payment_period_end: Optional[date] = None
    payment_method: Optional[str] = Field(None, max_length=50)
    
    notes: Optional[str] = None
    
    # Options
    calculate_taxes: bool = True  # Should we calculate taxes if not provided?
    update_next_payment: bool = True


class IncomeHistoryResponse(BaseModel):
    """Response for income history entry"""
    history_id: UUID
    income_id: UUID
    
    hours_worked: Optional[Decimal] = None
    days_worked: Optional[Decimal] = None
    sessions_worked: Optional[int] = None
    
    gross_amount: Decimal
    tax_federal: Decimal
    tax_state: Decimal
    tax_local: Decimal
    tax_fica: Decimal
    other_deductions: Decimal
    total_deductions: Decimal
    net_amount: Decimal
    
    expected_gross: Optional[Decimal] = None
    variance: Optional[Decimal] = None
    
    payment_date: date
    payment_period_start: Optional[date] = None
    payment_period_end: Optional[date] = None
    payment_method: Optional[str] = None
    
    ai_calculated: bool
    user_confirmed: bool
    user_corrected: bool
    
    effective_tax_rate: float
    
    notes: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


# ============================================
# RESPONSE SCHEMAS
# ============================================

class IncomeSourceResponse(BaseModel):
    """Full income source response"""
    income_id: UUID
    user_id: UUID
    
    source_name: str
    source_type: str
    employer_name: Optional[str] = None
    description: Optional[str] = None
    
    pay_structure: str
    pay_rate: Optional[Decimal] = None
    pay_unit: Optional[str] = None
    fixed_amount: Optional[Decimal] = None
    
    frequency: str
    next_payment_date: Optional[date] = None
    
    max_hours_per_week: Optional[int] = None
    expected_hours_per_period: Optional[Decimal] = None
    
    is_taxable: bool
    tax_rate_federal: Optional[Decimal] = None
    tax_rate_state: Optional[Decimal] = None
    tax_rate_local: Optional[Decimal] = None
    tax_rate_fica: Optional[Decimal] = None
    tax_withholding_type: str
    
    is_guaranteed: bool
    reliability_score: Decimal
    
    is_active: bool
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    
    notes: Optional[str] = None
    
    created_at: datetime
    updated_at: datetime
    
    # Computed
    total_tax_rate: float
    estimated_gross_per_period: float
    estimated_net_per_period: float
    estimated_monthly_gross: float
    estimated_monthly_net: float
    
    # Recent history
    recent_payments: List[IncomeHistoryResponse] = []

    class Config:
        from_attributes = True


class IncomeSourceSummary(BaseModel):
    """Lightweight income source for lists"""
    income_id: UUID
    source_name: str
    source_type: str
    employer_name: Optional[str] = None
    
    pay_structure: str
    frequency: str
    next_payment_date: Optional[date] = None
    
    estimated_monthly_net: float
    is_guaranteed: bool
    is_active: bool

    class Config:
        from_attributes = True


# ============================================
# STATISTICS
# ============================================

class IncomeStats(BaseModel):
    """Income statistics"""
    total_monthly_gross: Decimal
    total_monthly_net: Decimal
    total_annual_gross: Decimal
    total_annual_net: Decimal
    
    guaranteed_monthly: Decimal
    variable_monthly: Decimal
    
    active_sources_count: int
    
    average_effective_tax_rate: float
    
    by_type: dict  # {"job": 2000, "scholarship": 7900}
    by_source: dict  # {"IT Job": 1000, "Engineering Pathways": 500}


class MonthlyIncomeBreakdown(BaseModel):
    """Monthly income breakdown"""
    year: int
    month: int
    
    total_gross: Decimal
    total_net: Decimal
    total_taxes: Decimal
    
    by_source: dict
    
    vs_last_month: Optional[Decimal] = None
    vs_expected: Optional[Decimal] = None


# ============================================
# SPECIALIZED SCHEMAS
# ============================================

class StudentJobSetup(BaseModel):
    """Quick setup for a student job"""
    job_name: str
    employer_name: str
    hourly_rate: Decimal
    max_hours_per_week: int = 20
    expected_hours_per_week: Decimal
    frequency: PaymentFrequency = PaymentFrequency.BIWEEKLY
    start_date: Optional[date] = None
    
    # Simple tax setup
    federal_tax_rate: Decimal = Field(12.0, description="Federal tax bracket %")
    state_tax_rate: Decimal = Field(5.0, description="State tax %")


class ScholarshipSetup(BaseModel):
    """Quick setup for a scholarship"""
    scholarship_name: str
    institution_name: str
    semester_amount: Decimal
    covers: List[str] = ["meals", "rent", "utilities"]  # What it covers
    start_date: Optional[date] = None
    semesters_remaining: int = 1

File: frontend/src/api/types/income.ts
// src/api/types/income.ts
/**
 * Income Types - Aligned with Backend
 */

// ============ ENUMS ============

export type IncomeType = 'employment' | 'freelance' | 'business' | 'investment' | 'rental' | 'scholarship' | 'other';
export type PayStructure = 'salary' | 'hourly' | 'commission' | 'contract' | 'variable';
export type PayFrequency = 'weekly' | 'biweekly' | 'semimonthly' | 'monthly' | 'quarterly' | 'annually' | 'irregular';
export type IncomeStatus = 'active' | 'paused' | 'ended';
export type DeductionType = 'pre_tax' | 'post_tax';
export type HistoryStatus = 'confirmed' | 'pending' | 'expected';

// ============ DEDUCTION ============

export interface Deduction {
  deduction_id: string;
  deduction_name: string;
  deduction_type: DeductionType;
  amount: number;
  is_percentage: boolean;
}

export interface CreateDeductionRequest {
  deduction_name: string;
  deduction_type: DeductionType;
  amount: number;
  is_percentage: boolean;
}

// ============ INCOME SOURCE ============

export interface CreateIncomeSourceRequest {
  source_name: string;
  employer_name?: string;
  income_type: IncomeType;
  pay_structure: PayStructure;
  base_amount: number;
  hourly_rate?: number;
  expected_hours?: number;
  expected_hours_per_period?: number;
  pay_frequency: PayFrequency;
  next_payment_date?: string;
  pay_day?: number;
  max_hours_per_week?: number;
  is_taxable?: boolean;
  federal_tax_rate?: number;
  state_tax_rate?: number;
  local_tax_rate?: number;
  is_guaranteed?: boolean;
  deductions?: CreateDeductionRequest[];
  status?: IncomeStatus;
  start_date?: string;
  end_date?: string;
  notes?: string;
}

export interface UpdateIncomeSourceRequest {
  source_name?: string;
  employer_name?: string;
  income_type?: IncomeType;
  pay_structure?: PayStructure;
  base_amount?: number;
  hourly_rate?: number;
  expected_hours?: number;
  expected_hours_per_period?: number;
  pay_frequency?: PayFrequency;
  next_payment_date?: string;
  pay_day?: number;
  max_hours_per_week?: number;
  is_taxable?: boolean;
  federal_tax_rate?: number;
  state_tax_rate?: number;
  local_tax_rate?: number;
  is_guaranteed?: boolean;
  deductions?: CreateDeductionRequest[];
  status?: IncomeStatus;
  start_date?: string;
  end_date?: string;
  notes?: string;
}

export interface IncomeSource {
  income_id: string;
  source_name: string;
  employer_name?: string;
  // Type fields (backend sends both)
  income_type: IncomeType;
  source_type: IncomeType;
  pay_structure: PayStructure;
  // Amount fields (backend sends both)
  base_amount: number;
  amount: number;
  hourly_rate?: number;
  expected_hours?: number;
  // Frequency fields (backend sends both)
  pay_frequency: PayFrequency;
  frequency: PayFrequency;
  next_payment_date?: string;
  pay_day?: number;
  max_hours_per_week?: number;
  // Tax fields
  is_taxable: boolean;
  federal_tax_rate: number;
  state_tax_rate: number;
  local_tax_rate: number;
  fica_rate: number;
  total_tax_rate: number;
  // Computed values
  gross_per_period: number;
  net_per_period: number;
  taxes_per_period: number;
  monthly_gross: number;
  monthly_net: number;
  annual_gross: number;
  annual_net: number;
  // Status flags
  is_guaranteed: boolean;
  is_recurring: boolean;
  is_active: boolean;
  status: IncomeStatus;
  // Related
  deductions: Deduction[];
  // Dates
  start_date?: string;
  end_date?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface IncomeSourceSummary {
  income_id: string;
  source_name: string;
  source_type: IncomeType;
  pay_structure: PayStructure;
  amount: number;
  frequency: PayFrequency;
  next_payment_date?: string;
  is_active: boolean;
  monthly_net: number;
}

// ============ INCOME HISTORY ============

export interface LogIncomeRequest {
  gross_amount: number;
  net_amount?: number;
  payment_date: string;
  pay_period_start?: string;
  pay_period_end?: string;
  hours_worked?: number;
  tax_federal?: number;
  tax_state?: number;
  tax_local?: number;
  tax_fica?: number;
  notes?: string;
}

export interface IncomeHistoryEntry {
  history_id: string;
  income_id: string;
  gross_amount: number;
  net_amount: number;
  tax_federal: number;
  tax_state: number;
  tax_local: number;
  tax_fica: number;
  hours_worked?: number;
  payment_date: string;
  pay_period_start?: string;
  pay_period_end?: string;
  status: HistoryStatus;
  is_confirmed: boolean;
  is_recurring: boolean;
  notes?: string;
  created_at: string;
}

// ============ STATISTICS ============

export interface IncomeStats {
  total_monthly_gross: number;
  total_monthly_net: number;
  total_annual_gross: number;
  total_annual_net: number;
  guaranteed_monthly: number;
  variable_monthly: number;
  active_sources_count: number;
  average_tax_rate: number;
}